### PR TITLE
[New block] Post Navigation Link

### PR DESCRIPTION
**Pull request by ntsekouras**
## Description
This PR creates a new block `Post Navigation Link` that is used to display the previous or next post link, that is adjacent to the current post. It works with `pages`, `posts` and `custom post types`.

This base block can only be used through its block variations: `Next post link` and `Previous post link`. If not in `singular` page it returns nothing.

This block has an option to show the next/previous post's title and in general can produce three different results:
1. Show title is set to `false` and **no** custom label -> Link with the default text that is `Next` or `Previous` accordingly
2. Show title is set to `true` and **no** custom label -> A link with the post's title
3. Show title is set to `true` and a custom label -> A link with the post's title with the custom label prepended to it.

## How has this been tested?
I have tested with `pages`, `posts` and `custom post types` but wasn't able to properly test in a block-theme because of the recent problems that not load properly.

## Screenshots 
I will add gifs when the themes problem is resolved.

## Follow ups
1. Add support for showing links for the same term and handle taxonomy.

## Notes
There is no icon (example arrow icon) at the produced link, but it can be added with css by theme authors for now. Later we could consider a control for setting such an icon.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by carolinan** [comment]
Hi
I am confused by the naming, I think the s in posts needs to be removed?

The **post** navigation is what we now call the navigation between single posts
https://developer.wordpress.org/reference/functions/the_post_navigation/

The **posts** navigation is navigation between a set of posts, and in Gutenberg, this is the query pagination blocks?
https://developer.wordpress.org/reference/functions/the_posts_navigation/
(And https://developer.wordpress.org/reference/functions/the_posts_pagination/)

**Comment by ntsekouras** [comment]
> I am confused by the naming, I think the s in posts needs to be removed?

You're right. I renamed it.

**Comment by gziolo** [comment]
Shouldn't the codename of the block align with the title – `post-navigation-link`?

**Comment by ntsekouras** [comment]
>Shouldn't the codename of the block align with the title – post-navigation-link?

This makes even more sense. I believe this will be the last rename :)

**Review by gziolo — COMMENTED** [review]
Nice work. I really like how the block variations API simplifies so many things 👍🏻

**Comment by kristastevens** [comment]
Hello! The "Needs Copy Review" label was attached to this PR. Is there something specific you'd like us to look at?

**Comment by ntsekouras** [comment]
>Hello! The "Needs Copy Review" label was attached to this PR. Is there something specific you'd like us to look at?

Hey @kristastevens - labels and descriptions from the block and its variations like `Displays the next or previous post link that is adjacent to the current post.`. Most of them were copied with small adjustments from the WP documentation like: https://developer.wordpress.org/reference/functions/get_next_post_link/

**Comment by kristastevens** [comment]
@ntsekouras where are the labels and descriptions? I'm not seeing any copy that needs review. (Sorry! I don't mean to be difficult.)

**Comment by ntsekouras** [comment]
>where are the labels and descriptions

@kristastevens Here are the links:

https://github.com/WordPress/gutenberg/pull/28602/files#diff-e74a25529ee6f669153ea4d7b54f2d160f8227e87def564cc74b7eb29669125eR22

https://github.com/WordPress/gutenberg/pull/28602/files#diff-a16e40575336348a2ff08c0ef5c282976c37d60a89942bbd7b3ad72b8406c321R15

https://github.com/WordPress/gutenberg/pull/28602/files#diff-a16e40575336348a2ff08c0ef5c282976c37d60a89942bbd7b3ad72b8406c321R25


Did I need to add the `needs copy review` for them?

**Comment by kristastevens** [comment]
Thanks so much, @ntsekouras! Please always feel welcome to request copy review on anything you like. 

> 'Displays the next post link that is adjacent to the current post.'

I think we can trim this a tiny bit without compromising the meaning / intent. Suggestions: 

> 'Displays the post link that follows the current post.' 

> Displays the post link published after the current post.' 

Same idea here with a small trim:

> 'Displays the previous post link that is adjacent to the current post.' 

Suggestions: 
> 'Displays the post link that precedes the current post.'

> 'Displays the post link published before the current post.'

**Comment by mkaz** [comment]
One issue, I could not added the block when using a normal theme, when I switched to a the TT1 block-based theme it worked. I couldn't figure out why this was the case though.

A couple of usability bumps:

* I feel like when toggling "Display the title as a link" some sort of placeholder should be included in the editor. 

* I default expected an icon to show, might be a good starting state, the default of just typing a label confused me at first without any placeholder or guidance.

* My attempt at a basic two column, left and right links for previous and next, I added a two-column Column block and then added the links inside. For the right column, I want to align the item right but there is no option. Something like this might be a nice default pattern.

Functionality all seemed to work fine.

**Comment by ntsekouras** [comment]
Hey @mkaz - thanks for testing!

>I could not added the block when using a normal theme

This block is registered under the FSE umbrella.

>I default expected an icon to show, might be a good starting state, the default of just typing a label confused me at first without any placeholder or guidance.

Good point! I made the placeholder persist on empty value so it will provide more guidance.
Also about the icon as per the PR's description:

>There is no icon (example arrow icon) at the produced link, but it can be added with css by theme authors for now. Later we could consider a control for setting such an icon.

The key here is that we should give more thought about a control for setting `icons` as it will be used by other blocks as well.

>I feel like when toggling "Display the title as a link" some sort of placeholder should be included in the editor.

This is a tricky one and that's why in the toggle there is this hint: `If you have entered a custom label, it will be prepended before the title.`. I wanted to avoid the complexity of using format replacements like `%title` for now. We can revisit in the future and using such formats will be needed at least here: https://github.com/WordPress/gutenberg/pull/27989, for the `Search Title` block variation. This will require a lot of thought to make it look, work and feel well.

>My attempt at a basic two column, left and right links for previous and next, I added a two-column Column block and then added the links inside. For the right column, I want to align the item right but there is no option.

I'll look into that!

File: packages/icons/src/library/post-navigation-link-previous.js [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/**
+ * WordPress dependencies
+ */
+import { SVG, Path } from '@wordpress/primitives';
+
+const postNavigationLinkPrevious = (
```
Comment (mtias) [inline packages/icons/src/library/post-navigation-link-previous.js:6]: I'd simplify `post-previous`

File: lib/blocks.php [hunk 2]
```diff
@@ -76,6 +76,7 @@ function gutenberg_reregister_core_block_types() {
 					'post-excerpt.php'              => 'core/post-excerpt',
 					'post-featured-image.php'       => 'core/post-featured-image',
 					'post-hierarchical-terms.php'   => 'core/post-hierarchical-terms',
+					'post-navigation-link.php'      => 'core/post-navigation-link',
```
Comment (mtias) [inline lib/blocks.php:79]: Would `post-previous-next` be more clear?

File: packages/block-library/src/post-navigation-link/variations.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+import { __ } from '@wordpress/i18n';
+import {
+	postNavigationLinkNext,
+	postNavigationLinkPrevious,
+} from '@wordpress/icons';
+
+const variations = [
+	{
+		isDefault: true,
+		name: 'post-navigation-link-next',
```
Comment (mtias) [inline packages/block-library/src/post-navigation-link/variations.js:13]: I'd simplify to `post-next` and `post-previous`

File: packages/block-library/src/post-navigation-link/variations.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+		title: __( 'Next post link' ),
```
Comment (mtias) [inline packages/block-library/src/post-navigation-link/variations.js:14]: I think for block title "Next Post" might be better

File: lib/blocks.php (hunk 2 above)
Comment (ntsekouras) [reply lib/blocks.php:79]: I'm not really sure about that.. I've no strong opinion on this though.

**Review by mkaz — APPROVED** [review]
The placeholder update looks good. :+1: 

I missed the part about the icon, I agree that can be addressed later.

I didn't see the update for the alignment in columns, but I think that also can be addressed in other tickets and not unique to the nav links.

Looking good, worth landing and iterating based on further feedback. :rocket:

**Comment by ntsekouras** [comment]
Thanks for testing and reviewing @mkaz!

>I didn't see the update for the alignment in columns

I added `textAlignment` in these new blocks, not something in columns. Let's see how this goes :)

**Pull request merged**